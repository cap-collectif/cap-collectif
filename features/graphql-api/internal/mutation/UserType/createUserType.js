import '../../../_setup';

const CreateUserTypeMutation = /* GraphQL */ `
  mutation CreateUserTypesMutation($input: CreateUserTypeInput!) {
    createUserType(input: $input) {
      userType {
        id
        name
        media {
          id
        }
      }
    }
  }
`;

const UserTypesQuery = /* GraphQL */ `
  query {
    userTypes {
      id,
      name,
      media {
        id
      }
    }
  }
`;

const creationSnapshotPattern = {
  createUserType: {
    userType: {
      id: expect.any(String),
    },
  },
};

describe('mutations.createUserType', () => {
  let userTypeCount;
  let createdUserTypeIds = [];
  beforeAll(async () => {
    const result = await graphql(UserTypesQuery, null, 'internal_admin');
    expect(result).toMatchObject({ userTypes: expect.any(Array) });
    userTypeCount = result.userTypes.length;
  });

  it('creates a new user type', async () => {
    const result = await graphql(
      CreateUserTypeMutation,
      {
        input: {
          translations: [
            {
              name: 'ONG',
              locale: 'FR_FR',
            },
            {
              name: 'NGO',
              locale: 'EN_GB',
            },
          ],
        },
      },
      'internal_admin',
    );
    expect(result).toMatchSnapshot(creationSnapshotPattern);
    createdUserTypeIds.push(result.createUserType.userType.id);
  });

  // the user-type-nonprofit should be in the fixtures
  it('creates a new user type with a media', async () => {
    const result = await graphql(
        CreateUserTypeMutation,
        {
          input: {
            translations: [
              {
                name: 'Licorne',
                locale: 'FR_FR',
              },
            ],
            media: 'user-type-nonprofit',
          },
        },
        'internal_admin',
    );
    expect(result).toMatchSnapshot(creationSnapshotPattern);
    createdUserTypeIds.push(result.createUserType.userType.id);
  });

  it('checks user types after creation', async () => {
    const result = await graphql(UserTypesQuery, null, 'internal_admin');
    expect(result).toMatchObject({ userTypes: expect.any(Array) });
    expect(result.userTypes).toHaveLength(userTypeCount + 2);
    for (const createdId of createdUserTypeIds) {
      expect(result.userTypes.some(u => u.id === createdId)).toBe(true);
    }
  });

  it('creates a new user type with a non-existing media', async () => {
    await expect(
      graphql(
        CreateUserTypeMutation,
        {
          input: {
            translations: [
              {
                name: 'Boîte sans media',
                locale: 'FR_FR',
              },
            ],
            media: 'fake-media-id-for-user-type-test',
          },
        },
        'internal_admin',
      ),
    ).rejects.toThrowError('This value is not valid.');
  });

  it('tries to create a user type as a basic user', async () => {
    await expect(
        graphql(
            CreateUserTypeMutation,
            {
              input: {
                translations: [
                  {
                    name: 'G pas les droits !',
                    locale: 'FR_FR',
                  },
                ],
              },
            },
            'internal_user',
        ),
    ).rejects.toThrowError('Access denied to this field.');
  });

  it('tries to create a user type while being unauthenticated', async () => {
    await expect(
        graphql(
            CreateUserTypeMutation,
            {
              input: {
                translations: [
                  {
                    name: 'Je suis pas loggé',
                    locale: 'FR_FR',
                  },
                ],
              },
            },
            'internal',
        ),
    ).rejects.toThrowError('Access denied to this field.');
  });
});
