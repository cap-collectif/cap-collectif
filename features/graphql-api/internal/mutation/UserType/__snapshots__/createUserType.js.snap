// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutations.createUserType creates a new user type 1`] = `
Object {
  "createUserType": Object {
    "userType": Object {
      "id": Any<String>,
      "media": null,
      "name": "ONG",
    },
  },
}
`;

exports[`mutations.createUserType creates a new user type with a media 1`] = `
Object {
  "createUserType": Object {
    "userType": Object {
      "id": Any<String>,
      "media": Object {
        "id": "user-type-nonprofit",
      },
      "name": "Licorne",
    },
  },
}
`;
